File: packages/backend/test-openai-api.ts
import axios from 'axios';

const BASE_URL = 'http://localhost:3001';

async function testOpenAIConnection() {
  console.log('Testing OpenAI API Connection...\n');
  
  const results: any = {};
  
  try {
    // 1. Check if backend is accessible
    console.log('1. Checking if backend is accessible...');
    try {
      const response = await axios.get(`${BASE_URL}/api/health`);
      results.backendAccessible = true;
      results.healthStatus = response.status;
      console.log(`   ✓ Backend is accessible (status: ${response.status})`);
    } catch (error: any) {
      results.backendAccessible = false;
      console.log(`   ✗ Backend not accessible: ${error.message}`);
      return results;
    }
    
    // 2. Reset OpenAI circuit breaker
    console.log('\n2. Resetting OpenAI circuit breaker...');
    try {
      const response = await axios.post(`${BASE_URL}/api/v1/circuit-breaker/reset/openai-api`);
      results.circuitBreakerReset = response.status === 200;
      console.log(`   Circuit breaker reset: ${response.status}`);
      if (response.data) {
        console.log(`   Response:`, response.data);
      }
    } catch (error: any) {
      results.circuitBreakerReset = false;
      console.log(`   ✗ Failed to reset circuit breaker: ${error.message}`);
    }
    
    // 3. Check circuit breaker status
    console.log('\n3. Checking circuit breaker status...');
    try {
      const response = await axios.get(`${BASE_URL}/api/v1/circuit-breaker/status/openai-api`);
      results.circuitBreakerState = response.data.state || 'UNKNOWN';
      console.log(`   Circuit breaker state: ${response.data.state || 'UNKNOWN'}`);
      console.log(`   Full status:`, JSON.stringify(response.data, null, 2));
    } catch (error: any) {
      results.circuitBreakerState = 'ERROR';
      console.log(`   ✗ Failed to check status: ${error.message}`);
    }
    
    // 4. Test OpenAI connection
    console.log('\n4. Testing OpenAI connection...');
    try {
      const response = await axios.get(`${BASE_URL}/api/v1/sentiment/openai/test`);
      results.openaiTestSuccess = response.status === 200;
      console.log(`   OpenAI test status: ${response.status}`);
      console.log(`   Response:`, JSON.stringify(response.data, null, 2));
    } catch (error: any) {
      results.openaiTestSuccess = false;
      console.log(`   ✗ OpenAI test failed: ${error.message}`);
      if (error.response?.data) {
        console.log(`   Error details:`, error.response.data);
      }
    }
    
    // 5. Test sentiment analysis
    console.log('\n5. Testing sentiment analysis...');
    try {
      const response = await axios.post(`${BASE_URL}/api/v1/sentiment/analyze`, {
        text: "This is a wonderful test!",
        provider: "openai",
        model: "gpt-3.5-turbo"
      });
      results.sentimentAnalysisSuccess = response.status === 200;
      console.log(`   Sentiment analysis status: ${response.status}`);
      console.log(`   Response:`, JSON.stringify(response.data, null, 2));
    } catch (error: any) {
      results.sentimentAnalysisSuccess = false;
      console.log(`   ✗ Sentiment analysis failed: ${error.message}`);
      if (error.response?.data) {
        console.log(`   Error details:`, error.response.data);
      }
    }
    
    // 6. Check final circuit breaker status
    console.log('\n6. Checking final circuit breaker status...');
    try {
      const response = await axios.get(`${BASE_URL}/api/v1/circuit-breaker/status/openai-api`);
      results.finalCircuitBreakerState = response.data.state || 'UNKNOWN';
      console.log(`   Final circuit breaker state: ${response.data.state || 'UNKNOWN'}`);
      console.log(`   Full status:`, JSON.stringify(response.data, null, 2));
    } catch (error: any) {
      results.finalCircuitBreakerState = 'ERROR';
      console.log(`   ✗ Failed to check final status: ${error.message}`);
    }
    
    // Summary
    console.log('\n' + '='.repeat(50));
    console.log('SUMMARY:');
    console.log('='.repeat(50));
    console.log(`Backend accessible: ${results.backendAccessible || false}`);
    console.log(`Circuit breaker state: ${results.circuitBreakerState || 'UNKNOWN'} → ${results.finalCircuitBreakerState || 'UNKNOWN'}`);
    console.log(`OpenAI test successful: ${results.openaiTestSuccess || false}`);
    console.log(`Sentiment analysis successful: ${results.sentimentAnalysisSuccess || false}`);
    
    return results;
    
  } catch (error: any) {
    console.error('Unexpected error:', error);
    return results;
  }
}

// Run the test
testOpenAIConnection().then(() => {
  process.exit(0);
}).catch((error) => {
  console.error('Test failed:', error);
  process.exit(1);
});

File: packages/backend/test_openai_api.py
#!/usr/bin/env python3
import urllib.request
import urllib.parse
import json
import sys

print("=== OpenAI API Test ===\n")

base_url = "http://localhost:3001"

try:
    # Test 1: Reset Circuit Breaker
    print("1. Resetting circuit breaker...")
    reset_req = urllib.request.Request(
        f"{base_url}/api/v1/circuit-breaker/reset/openai-api",
        method='POST'
    )
    try:
        with urllib.request.urlopen(reset_req) as response:
            data = json.loads(response.read().decode())
            print(f"   ✓ Status: {response.status}")
            print(f"   ✓ Response: {data}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
    
    print("\n2. Checking circuit breaker status...")
    try:
        with urllib.request.urlopen(f"{base_url}/api/v1/circuit-breaker/status/openai-api") as response:
            data = json.loads(response.read().decode())
            metrics = data.get('metrics', {})
            print(f"   ✓ State: {metrics.get('state', 'Unknown')}")
            print(f"   ✓ Success Count: {metrics.get('successCount', 0)}")
            print(f"   ✓ Failure Count: {metrics.get('failureCount', 0)}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
    
    print("\n3. Testing OpenAI connection...")
    try:
        with urllib.request.urlopen(f"{base_url}/api/v1/sentiment/openai/test") as response:
            data = json.loads(response.read().decode())
            print(f"   ✓ Status: {response.status}")
            print(f"   ✓ Response: {data}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
    
    print("\n4. Testing sentiment analysis...")
    analysis_data = json.dumps({
        "text": "This product is absolutely amazing!",
        "provider": "openai",
        "model": "gpt-3.5-turbo"
    }).encode('utf-8')
    
    analysis_req = urllib.request.Request(
        f"{base_url}/api/v1/sentiment/analyze",
        data=analysis_data,
        headers={'Content-Type': 'application/json'},
        method='POST'
    )
    
    try:
        with urllib.request.urlopen(analysis_req) as response:
            result = json.loads(response.read().decode())
            if 'sentiment' in result:
                print(f"   ✅ SUCCESS! Analysis completed")
                print(f"   ✓ Sentiment: {result['sentiment']}")
                print(f"   ✓ Score: {result.get('score', 'N/A')}")
                print(f"   ✓ Confidence: {result.get('confidence', 'N/A')}")
            else:
                print(f"   ❌ FAILED: {result}")
    except urllib.error.HTTPError as e:
        error_data = e.read().decode()
        print(f"   ❌ HTTP Error {e.code}: {error_data}")
    except Exception as e:
        print(f"   ❌ Failed: {e}")
    
    print("\n5. Final circuit breaker check...")
    try:
        with urllib.request.urlopen(f"{base_url}/api/v1/circuit-breaker/status/openai-api") as response:
            data = json.loads(response.read().decode())
            metrics = data.get('metrics', {})
            state = metrics.get('state', 'Unknown')
            print(f"   ✓ Final State: {state}")
            
            if state == 'CLOSED':
                print("\n✅ SUCCESS: OpenAI API is working correctly!")
            else:
                print(f"\n⚠️  WARNING: Circuit breaker is {state}")
    except Exception as e:
        print(f"   ✗ Failed: {e}")
        
except Exception as e:
    print(f"\n❌ Backend is not accessible: {e}")
    print("Please ensure the backend is running on port 3001")
    sys.exit(1)

print("\n=== Test Complete ===")